singer/src/main/java/com/pinterest/singer/writer/KafkaWriter.java

```diff
@@ -332,13 +342,85 @@ public ExecutorService getClusterThreadPool() {
     return buckets;
   }
 
+  /**
+   *  Validate the message and inject headers for the ProducerRecord
+   *
+   *  If the message is corrupted and corrupted messages are configured to be skipped at the current stage, return true.
+   *  Otherwise, return false, since the message should not be skipped
+   *
+   * @param msg the message to validate
+   * @param headers the headers of the ProducerRecord
+   * @param indexWithinTheBucket the index of the current message within the bucket
+   * @param partitionId the partition id
+   * @param mapOfHeadersMaps a map of index
+   * @param mapOfMessageValidMaps
+   * @return
+   */
+  public boolean checkMessageValidAndInjectHeaders(
+          LogMessage msg, Headers headers, int indexWithinTheBucket, int partitionId,
+          Map<Integer, Map<Integer, LoggingAuditHeaders>> mapOfHeadersMaps,
+          Map<Integer, Map<Integer, Boolean>> mapOfMessageValidMaps) {
+    boolean shouldSkipMessage = false;
+    boolean isMessageValid = checkMessageValid(msg);
+    if (enableLoggingAudit && auditConfig.isSkipCorruptedMessageAtCurrentStage() && !isMessageValid) {
+      OpenTsdbMetricConverter.incr(SingerMetrics.NUM_CORRUPTED_MESSAGES_SKIPPED, "topic=" + topic,
+              "host=" + HOSTNAME, "logName=" + msg.getLoggingAuditHeaders().getLogName(),
+              "logStreamName=" + logName);
+      shouldSkipMessage = true;
+    }
+    if (this.headersInjector != null) {
+      injectHeadersForProducerRecord(msg, headers);
```

Comment:
@zzhhhzz is there a reason for still performing header injection if the message is corrupt & will be skipped?

Reply:
@ambud  yes.   When messages are decided as corrupted when the crc checksum computed by Singer is different from the value computed by thrift-logger,  we still want to injector headers as well as the messageCRC, because we can configure downstream system like Merced to detect and remove the corrupted messages.  the audit events sent out from different stages have LoggingAuditHeaders as one of its field and real-time processing will group audit events based on the LoggingAuditHeaders to decide data loss & corruption.

Reply:
when shouldSkipMessage is true,  there is no need to inject headers.   should change the if conditions:    if (!shouldSkipMessage && this.headersInjector != null) ...